### ISSUE TITLE
'top' crashes with bus error

### ISSUE THREAD

**Issue by mikrosk** [freemint/freemint#325]
As reported here: https://sourceforge.net/p/freemint/mailman/message/58710892

The offending commit is 701612b5, @th-otto do you want to take a look? There's quite a few changes in that commit of yours.

**Comment by th-otto** [comment]
What do   you mean by offending? I don't get any crashes here, using the same binary.

**Comment by mikrosk** [comment]
Hmm, all I do is to download latest aranym/freemint snapshot, open toswin2 and run the `top`. It shows weird characters (first bug) and after a few seconds it ends up with bus error (second bug). Run with MP of course.

**Comment by th-otto** [comment]
Can you just download the sources and recompile them, and see if the bug still occurs?

**Comment by mikrosk** [comment]
I can't because there are actually no sources provided. :-)

`ps` looks ok (no weird characters nor crashing)

**Comment by th-otto** [comment]
I wonder what they were compiled with. Must be some really old gcc & mintlib, since they are really small, and don't even have a MiNT header.

**Comment by mikrosk** [comment]
Who knows. However since those are quite widespread (EasyMiNT...) this should be fixed (reverted in the worst case).

**Comment by th-otto** [comment]
The fix was there because the pseudo directories did not return "." and ".." entries. Without the patch, other tools like find will crash.

**Comment by skaftetryne** [comment]
> I wonder what they were compiled with. Must be some really old gcc & mintlib, since they are really small, and don't even have a MiNT header.

I believe most SpareMiNT packages were compiled with 2.95 on real hardware.

**Comment by skaftetryne** [comment]
> The fix was there because the pseudo directories did not return "." and ".." entries. Without the patch, other tools like find will crash.

find seems to work fine here, how can I reproduce this crash?

**Comment by th-otto** [comment]
Yeah, but i still wonder. The ps binary is ~26k and even includes a symbol table. If you strip it, it is ~18k. That is *really* small, even for old gcc and old mintlib.
```
$Patchlevel: MiNT library: 45
```

But maybe we can find some old sources elsewhere:
```
@(#)clrtoeol.c  5.3 (Berkeley) 6/30/88
@(#)cr_put.c    5.3 (Berkeley) 6/30/88
@(#)curses.c    5.5 (Berkeley) 6/30/88
@(#)endwin.c    5.3 (Berkeley) 6/30/88
@(#)erase.c     5.3 (Berkeley) 6/30/88
@(#)getch.c     5.5 (Berkeley) 6/30/88
@(#)getstr.c    5.3 (Berkeley) 6/30/88
@(#)initscr.c   5.3 (Berkeley) 6/30/88
@(#)move.c      5.4 (Berkeley) 6/30/88
@(#)newwin.c    5.3 (Berkeley) 6/30/88
@(#)printw.c    5.4 (Berkeley) 6/30/88
@(#)putchar.c   5.3 (Berkeley) 6/30/88
@(#)refresh.c   5.3 (Berkeley) 6/30/88
@(#)scroll.c    5.3 (Berkeley) 6/30/88
@(#)touchwin.c  5.3 (Berkeley) 6/30/88
@(#)tstp.c      5.3 (Berkeley) 6/30/88
$PatchLevel: atariST/TT Curses/Termcap: 16 $
@(#)addch.c     5.4 (Berkeley) 6/30/88
@(#)addstr.c    5.4 (Berkeley) 6/30/88
@(#)cr_tty.c    5.4 (Berkeley) 6/30/88
@(#)deleteln.c  5.3 (Berkeley) 6/30/88
@(#)delwin.c    5.3 (Berkeley) 6/30/88
@(#)id_subwins.c        5.3 (Berkeley) 6/30/88
@(#)longname.c  5.3 (Berkeley) 6/30/88
@(#)addbytes.c  5.3 (Berkeley) 6/30/88
```

**Comment by th-otto** [comment]
> find seems to work fine here, how can I reproduce this crash?

It only used to have some problems when running on pseudo filesystems like /proc, /kern etc.

**Comment by skaftetryne** [comment]
> Yeah, but i still wonder. The ps binary is ~26k and even includes a symbol table. If you strip it, it is ~18k. That is _really_ small, even for old gcc and old mintlib.

They are taken from KGMD:

https://ftp.funet.fi/pub/atari/mint/KGMD/

Compiled by Knarf in 1995.

**Comment by skaftetryne** [comment]
As the sources for the old MiNTOS top is not available I wrote my own:

https://atari.joska.no/jtop/

**Comment by mikrosk** [comment]
Not sure what are the chances that those are the same tools but found this one, directly from Atari's Allan Pratt.

[pstopls.zip](https://github.com/user-attachments/files/18172419/pstopls.zip)

**Comment by skaftetryne** [comment]
They could be the same, or at least based on the same sources. The binaries does not work correctly - ps reports 000 for all ppid/ppids and top crashes with a bus error.

**Comment by skaftetryne** [comment]
I wrote a clone of the MiNTOS "ps" instead of messing with the sources above:

https://atari.joska.no/ps/

Source included.

**Comment by th-otto** [comment]
Nice. Should we add this to the tools directory, and build it for the snaphots? However that would probably mean the binary will be larger than your Pure-C compiled program.

**Comment by skaftetryne** [comment]
Yes, you can add it to the repository if you want to, I'm not going to do more with this code myself. I have not tried compiling it with gcc/MiNTlib so some changes are probably required. And yes, I expect the code to grow beyond 5.5Kb :D It doesn't need anything special though, so maybe it could be compiled with libcmini to make it more suitable for systems with "normal" amounts of RAM.

**Comment by skaftetryne** [comment]
Turns out I wasn't quite finished yet... Updated version can be found at https://atari.joska.no/ps/

The bug mentioned by @th-otto has been fixed, and a workaround for #394 has been added.

**Comment by th-otto** [comment]
Hm, interesting. Do you have a similar workaround in jtop? Because there, pid 0 and 1 appear as usual, with names MiNT and init.

**Comment by th-otto** [comment]
BTW, the executable compiled with libcmini is ~8k, compared to ~6k of your version. I think that is acceptable.

**Comment by th-otto** [comment]
Added now to the repo, with a few changes:
- struct ploadinfo was declared with "int" for fnamelen. That works for Pure-C, but not for gcc. Changed to short.
- struct XATTR is not declared in mintlib headers. This is already a problem in other projects. In libcmini it is declared, but with different member names than yours. Added a declaration for this (looks a bit strange, because i only renamed the fields that you actually use, attr & size)
- A minor tweak to return with exit code 1 on errors.

**Comment by skaftetryne** [comment]
Yes, there's a similar workaround in jtop.On 21 Dec 2024 05:45, Thorsten Otto ***@***.***> wrote:
Hm, interesting. Do you have a similar workaround in jtop? Because there, pid 0 and 1 appear as usual, with names MiNT and init.

**Comment by th-otto** [comment]
BTW, would be nice to get the same tools also working on MagiC. Currently, ps aborts with
```
Fcntl failed: -32
```
That is most likely due to Fcntl(PPROCADDR) not being supported. But since we have the MagiC sources, that could be fixed ;)

jtop does not error out, but does not print much either:
![Screenshot_20250101_133633](https://github.com/user-attachments/assets/ee3925ba-f3a6-42ec-9b1a-e2ee0286e578)

Have to check what jtop needs (most likely same as ps i guess). I'm also not sure whether vt52.prg supports colors, but that could also be fixed if needed.

**Comment by skaftetryne** [comment]
Yes, jtop does exactly the same when collecting information about processes. In addition it depends on /kern for the system info at the top (CPU info and memory).

jtop only use colours when the COLORTERM variable indicates that colours are available, so it shouldn't matter if vt52.prg supports colour or not.

**Comment by mikrosk** [comment]
> Yes, colors seem to be supported

I had been a MagiC user for years but this is the first time I ever see a colour output in VT52. Did you have to set something special?

**Comment by th-otto** [comment]
for jtop: set COLORTERM=1

I think there aren't many tos programs that use VT52 colors, therefor i wasn't sure whether VT52.PRG supported colors. Also you usually don't use programs on MagiC that were compiled with ncurses, and need a terminfo database.

But it is surely possible to write a simple test program that prints the color sequences and see what happens.

**Comment by th-otto** [comment]
Update: just try the attached program:
[vt52test.zip](https://github.com/user-attachments/files/18290345/vt52test.zip)

Note that it will not work in toswin2 with tw100 emulation. Even with tw52 emulation it prints different colors than in magic (or singletos), because the tw52 emulation seems to use a different color mapping. Normally, the VT52 emulation in tos just uses the color as the pixel value for 16 color resolutions.

**Comment by skaftetryne** [comment]
I reported the TosWin2 colour issue a few weeks ago here: https://github.com/freemint/toswin2/issues/2

**Comment by skaftetryne** [comment]
> BTW, when i press "key down" in magic, the program aborts with "ABNORMAL PROGRAM TERMINATION"

Yes, there's an assert that kicks in when the internal process list is empty (which should never happen).

**Comment by skaftetryne** [comment]
> So the vt52.prg really uses "red" for the text, but in toswin2 the text is dark blue.

I attempt to map the TW-colours to VT52-colours, so jtop will actually show red text when instructed to do so. If it shows blue text in the sysinfo part under TosWin2 it's because it reads a different configuration file. Try to specifically use the same config file as under MagiC with 'jtop -c <filename>' to be sure.

Btw I'm currently doing a lot of testing and debugging under N.AES of the stuff I've coded for the last almost 20 years, as I've used XaAES exclusively since then. One thing I've noticed is that N.AES prevents signals from aborting a pending Cnecin, which again effectively prevents jtop from working as it uses Talarm to interrupt Cnecin at the specified intervals to update the process list.

**Comment by mikrosk** [comment]
> I'm not able to get N.AES working with current kernels.

Not sure how old/new kernels but others have reported N.AES working perfectly fine again with latest snapshots. There have been a couple of N.AES bugfixes during the past 2-3 moths.

**Comment by skaftetryne** [comment]
This is with the latest snapshot. The N.AES setup that works with 1.16 crashes during boot here. I have not looked into this further, will do that when time (and motivation) allows for it.

**Comment by MedourMehdi** [comment]
Few month ago I started to build procps (with quick and dirty patches). Don't remember if it was from procps or procps-ng. pscommand seems to work except for "aux" flags. top didn't work. I can search for a diff if needed.
[procps_temp.zip](https://github.com/user-attachments/files/18379394/procps_temp.zip)

**Comment by th-otto** [comment]
I prefer the much simpler version from joska. procps does too many linux specific things, and depends on the text output in the /proc directory.

**Comment by MedourMehdi** [comment]
Yes for sure, joska version is great

**Comment by th-otto** [comment]
BTW, does jtop Dxreaddir() or Dreaddir()? I think the status field (Wait, I/O, TSR etc.) is only available through the XATTR structure, but not through struct stat.

**Comment by th-otto** [comment]
Ah ok, thanks. I'm asking, because i thought i could get rid of the duplicate code in https://github.com/freemint/freemint/blob/c6dc8f039e4a1bcff38e94dad9876213c7032e40/sys/procfs.c#L273. But then Fxattr would fall back to build the XATTR structure from the STAT structure, and there only the readonly attribute is set: https://github.com/freemint/freemint/blob/c6dc8f039e4a1bcff38e94dad9876213c7032e40/sys/xfs_xdd.c#L76-L77, not the other flags because they cannot be derived from the unix mode.

**Comment by mikrosk** [comment]
Not sure what fixed it but I just booted EasyMiNT in vconsd, launched `top` and it almost works as expected, there is just one additional weird process "?????" with size 0 listed. Well, it doesn't crash, we have a replacement, so closing.

**Comment by skaftetryne** [comment]
That weird process is probably either . or .. that Thorsten added.
